pandas/core/arrays/categorical.py

```diff
@@ -400,20 +400,38 @@ def astype(self, dtype: Dtype, copy: bool = True) -> ArrayLike:
             If copy is set to False and dtype is categorical, the original
             object is returned.
         """
-        if is_categorical_dtype(dtype):
+        if self.dtype is dtype:
+            result = self.copy() if copy else self
+
+        elif is_categorical_dtype(dtype):
             dtype = cast(Union[str, CategoricalDtype], dtype)
 
             # GH 10696/18593
             dtype = self.dtype.update_dtype(dtype)
             self = self.copy() if copy else self
-            if dtype == self.dtype:
-                return self
-            return self._set_dtype(dtype)
-        if is_extension_array_dtype(dtype):
-            return array(self, dtype=dtype, copy=copy)
-        if is_integer_dtype(dtype) and self.isna().any():
+            result = self._set_dtype(dtype)
+
+        elif is_extension_array_dtype(dtype):
+            result = array(self, dtype=dtype, copy=copy)
+
+        elif is_integer_dtype(dtype) and self.isna().any():
             raise ValueError("Cannot convert float NaN to integer")
-        return np.array(self, dtype=dtype, copy=copy)
+
+        elif len(self.codes) == 0 or len(self.categories) == 0:
+            result = np.array(self, dtype=dtype, copy=copy)
+
+        else:
+            # GH8628 (PERF): astype category codes instead of astyping array
+            try:
+                astyped_categories = self.categories.astype(dtype=dtype, copy=copy)
+            except (TypeError, ValueError):
+                raise ValueError(
+                    f"Cannot cast {self.categories.dtype} dtype to {dtype}"
+                )
+            astyped_categories = extract_array(astyped_categories, extract_numpy=True)
+            result = np.array(take_1d(astyped_categories, self._codes), copy=copy)
```

Comment:
you need to do `ensure_platform_int(self._code)` here to avoid 32-bit failing.

Reply:
Thanks!!! Added

Reply:
Hmmm I think I didn't do this right - 32 bit build still failing
```
(...)
        result = cat.astype(int)
        expected = np.array(cat, dtype=int)
>       tm.assert_numpy_array_equal(result, expected)
E       AssertionError: numpy array are different
E       
E       Attribute "dtype" are different
E       [left]:  int64
E       [right]: int32

```

Reply:
maybe pass `dtype=self._codes.dtype` to take_1d?

Reply:
Tried that, take_1d doesn't take a dtype arg...

Reply:
sorry, i meant to np.array

Reply:
Done (CI green now)